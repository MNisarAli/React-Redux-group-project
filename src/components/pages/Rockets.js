import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets, reserveRocket, cancelReservation } from '../../redux/rockets/rockets';

const Rockets = () => {
  // useDispatch hook to access the dispatch function from store.
  const dispatch = useDispatch();

  // useSelector hook to get rockets data from the store.
  const rockets = useSelector((state) => state.rockets);

  // `handleReserveRocket` dispatches `reserveRocket` action with selected rocket ID.
  const handleReserveRocket = (id) => {
    dispatch(reserveRocket(id));
  };

  // `handleCancelReservation` dispatches `cancelReservation` action with selected rocket ID.
  const handleCancelReservation = (id) => {
    dispatch(cancelReservation(id));
  };

  // useEffect hook to dispatch the `fetchRockets` action when the component is mounted,
  // to fetch the list of rockets from the API.
  useEffect(() => {
    dispatch(fetchRockets());
  });

  return (
    <section className="rockets-container">
      {rockets.map((rocket) => (
        <li className="rocket" key={rocket.id}>
          <img className="rocket-img" src={rocket.image} alt={rocket.name} />
          <article className="rocket-info">
            <h2 className="rocket-name">{rocket.name}</h2>
            <p className="rocket-description">
              {rocket.reserved && <span className="rocket-reserve-tag">Reserved</span>}
              {rocket.description}
            </p>
            {rocket.reserved ? (
              <button className="rocket-reserve-btn reserved" type="button" onClick={() => handleCancelReservation(rocket.id)}>
                Cancel Reservation
              </button>
            ) : (
              <button className="rocket-reserve-btn" type="button" onClick={() => handleReserveRocket(rocket.id)}>
                Reserve Rocket
              </button>
            )}
          </article>
        </li>
      ))}
    </section>
  );
};

export default Rockets;
