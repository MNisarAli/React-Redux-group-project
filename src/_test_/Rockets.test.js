import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Rockets from '../components/pages/Rockets';

const RocketProvider = () => (
  <Provider store={store}>
    <Rockets />
  </Provider>
);

describe('Rockets component testing', () => {
  it('checks for accurate rendering of all rockets', () => {
    const allRockets = renderer.create(<RocketProvider />).toJSON();
    expect(allRockets).toMatchSnapshot();
  });

  it('verifies the presence of the rockets-container text in the DOM', () => {
    const { container } = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
    expect(container.getElementsByClassName('rockets-container').length).toBe(1);
  });
});
