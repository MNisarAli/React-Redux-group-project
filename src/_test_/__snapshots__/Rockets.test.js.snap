// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component testing checks for accurate rendering of all rockets 1`] = `
<section
  className="rockets-container"
/>
`;
